modules/core_project_factory/scripts/preconditions.sh

```diff
@@ -14,17 +14,50 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# extract any --pip3_extra_flags from the incoming args, separating
+# them to their own array, to be passed to pip3 prior to calling
+# the precondition script.
+PIP_FLAGS=()
+REMAINING_FLAGS=()
+
+function extract_pip_flags {
+    local extract=0
+    local flag_re='--.*'
+
+    for flag do
+        shift
+        if [ "$flag" == "--pip3_extra_flags" ]; then
+            extract=1
+            continue
+        elif [ $extract -eq 1 ]; then
+            if ! [[ "$flag" =~ $flag_re ]]; then
+                if [ ${#flag} -gt 0 ]; then
+                    PIP_FLAGS+=("$flag")
+                fi
+            fi
+        fi
+        extract=0
+        REMAINING_FLAGS+=("$flag")
+    done
+}
+
 if command -v python3 1>/dev/null; then
     BASEDIR="$(dirname "$0")"
 
+    extract_pip_flags "$@"
+
     if command -v pip3 1>/dev/null; then
-        exec "pip3" "install" "-r" "$BASEDIR/preconditions/requirements.txt"
+        if [ "${#PIP_FLAGS[*]}" == "0" ]; then
```

Comment:
Technically, this conditional isn't actually required.  Because the linter wants all variables to be quoted, and sending an empty string as one of the params to exec does cause the empty param to be received by pip, I can't make it work correctly and pass lint without this extra conditional.

Technically, this is much more correct:

`exec "pip3" "install" ${PIP_FLAGS[*]} "-r" "path/to/requirements.txt"`

Since each element of the array will be sent separately.  But the linter wants to force quotes around the PIP_FLAGS reference.  If I put it in quotes, then pip objects to an empty arg being received.  So I added a conditional to prevent it from being sent at all unless it actually has a value in it, otherwise it is included, within a quoted string.

Reply:
We could also simply disable the linter check on this line if you think that'd improve the readability. Really, this seems a bit overcomplicated for simply injecting an extra flag—I'm surprised we need an `extract_pip_flags` flag at all.

One possible alternative/simpler implementation would be to only expose a "install_preconditions_as_user" variable which adds the required flag.

Reply:
Sure, but the difficulty is in how to actually pass that information to the preconditions.sh script.  If it goes in the args, then you have to get it back out before they are passed to the python script.  Otherwise, you need either a positional param we can shift off the front, or 2 versions of the script.  I'm happy to implement it either way, if someone wants to express a preference.  

I agree that this is a complex way to solve the problem, but it does keep duplicated code to a minimum and results in minimum impact to the user of the module while delivering the flexibility to pass arbitrary params to pip, rather than only allowing them to select between 2 options.  It is easy to imagine a requirement for a fair number of available options, especially for enterprise users, who can be stuck behind firewalls and internal artifact repositories, etc.

These are just the params that stand out as likely candidates for at least some users of the module.

```
  --user                      Install to the Python user install directory for your platform. Typically ~/.local/, or %APPDATA%\Python on Windows. (See the Python documentation for site.USER_BASE for full details.)
  --proxy <proxy>             Specify a proxy in the form [user:passwd@]proxy.server:port.
  --trusted-host <hostname>   Mark this host as trusted, even though it does not have valid or any HTTPS.
  --cert <path>               Path to alternate CA bundle.
  --client-cert <path>        Path to SSL client certificate, a single file containing the private key and the certificate in PEM format.
```

I did realize the PR as it stands right now, has a bug in it which will actually break if you set pip3_extra_flags to be '--user', anyway, since it will recognize that as a flag intended for the python script instead of pulling it into the pip flags.  I know how to fix it, but I'll wait until I hear a solution preference.

Reply:
What if we simple separated this into two different `local-exec` scripts? One could handle installing the pip dependencies, while the second actually executes the Python script.